### PR TITLE
Add Support for unordered_map (C++11) in AddressCache and GUID_t map objects in RtpsUdpDatalink

### PR DESCRIPTION
**Pull request by simpsont-oci**

### CODE REVIEW

File: dds/DCPS/Hash.h [hunk 1]
```diff
@@ -19,8 +23,40 @@ typedef unsigned char MD5Result[16];
...
+  return hash;
+}
+#endif
+
 }
 }
 OPENDDS_END_VERSIONED_NAMESPACE_DECL
 
+#ifdef ACE_HAS_CPP11
+#define OOAT_STD_HASH(Key, Export) \
```
Comment (iguessthislldo) [inline dds/DCPS/Hash.h:49]: Please start macros in headers with `OPENDDS_`.

File: dds/DCPS/Hash.cpp [hunk 2]
```diff
@@ -6,6 +6,7 @@
 #include "DCPS/DdsDcps_pch.h" //Only the _pch include should start with DCPS/
 
 #include "Hash.h"
+#include "Definitions.h"
```
Comment (mitza-oci) [inline dds/DCPS/Hash.cpp:9]: Is this needed?

File: dds/DCPS/AddressCache.h [hunk 3]
```diff
@@ -52,7 +53,15 @@ template <typename Key>
 class AddressCache {
 public:
 
+#if defined ACE_HAS_CPP11
+  typedef OPENDDS_UNORDERED_MAP_T(Key, RcHandle<AddressCacheEntry>) MapType;
+  typedef std::vector<Key> KeyVec;
```
Comment (mitza-oci) [inline dds/DCPS/AddressCache.h:58]: Use allocator-enabled vector (also below on 62)

File: dds/DCPS/AddressCache.h [hunk 4]
```diff
@@ -133,6 +152,7 @@ class AddressCache {
     return map_.erase(key) != 0;
   }
 
+  /*
```
Comment (mitza-oci) [inline dds/DCPS/AddressCache.h:155]: remove this method if it's not used

File: dds/DCPS/transport/rtps_udp/BundlingCacheKey.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  BundlingCacheKey(const GUID_t& dst_guid, const GUID_t& from_guid, const GuidSet& to_guids)
+    : dst_guid_(dst_guid)
+    , from_guid_(from_guid)
+    , to_guids_(to_guids)
+  {
+  }
+
+  bool operator<(const BundlingCacheKey& rhs) const
+  {
+    int r = std::memcmp(&dst_guid_, &rhs.dst_guid_, 2 * sizeof (GUID_t));
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/BundlingCacheKey.h:39]: This appears to be technically undefined behavior (but we're in uncharted waters with `#pragma pack`) and Coverity will likely complain.  At least change it to use the same kind of `memcmp` that `operator==` has, based on `this`.

File: dds/DCPS/transport/rtps_udp/BundlingCacheKey.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  {
+    if (this != &rhs) {
+      const_cast<GUID_t&>(dst_guid_) = rhs.dst_guid_;
+      const_cast<GUID_t&>(from_guid_) = rhs.from_guid_;
+      const_cast<GuidSet&>(to_guids_) = rhs.to_guids_;
+    }
+    return *this;
+  }
+
+  void contains(GuidSet& set) const
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/BundlingCacheKey.h:63]: The name `contains` reads like a predicate -- it wasn't clear what this was doing when I saw it used in AddressCache.h

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/Hash.cpp (hunk 2 above)
Comment (simpsont-oci) [reply dds/DCPS/Hash.cpp]: At one point, but not anymore... removing.

File: dds/DCPS/AddressCache.h (hunk 4 above)
Comment (simpsont-oci) [reply dds/DCPS/AddressCache.h]: Yep, gone.

File: dds/DCPS/AddressCache.h (hunk 3 above)
Comment (simpsont-oci) [reply dds/DCPS/AddressCache.h]: At one point, I wasn't supporting safety profile here, but obviously not the case anymore.

File: dds/DCPS/transport/rtps_udp/BundlingCacheKey.h (hunk 6 above)
Comment (simpsont-oci) [reply dds/DCPS/transport/rtps_udp/BundlingCacheKey.h]: changing to get_contained_guids

File: dds/DCPS/transport/rtps_udp/BundlingCacheKey.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  BundlingCacheKey(const GUID_t& dst_guid, const GUID_t& from_guid, const GuidSet& to_guids)
+    : dst_guid_(dst_guid)
+    , from_guid_(from_guid)
+    , to_guids_(to_guids)
+  {
+  }
+
+  bool operator<(const BundlingCacheKey& rhs) const
+  {
+    int r = std::memcmp(this, &rhs.dst_guid_, 2 * sizeof (GUID_t));
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/BundlingCacheKey.h:39]:
```suggestion
    int r = std::memcmp(this, &rhs, 2 * sizeof (GUID_t));
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**